Reorganize content into tabs of controller preferences

JoergAtGithub opened the pull request:
- Convert groupBoxSettings/groupBoxScreens into tabs
- Show only the tabs, which have valid content in the current context
- Sort tabs that settingsTab is shown by default, if controller settings are defined in mapping
- Moved controller tabs below info and warning section
- Made hyperlink warning label clickable and warning text select- and copy-able
- Made warning label collapsing in height
- Hide 'Mapping Info' fields if there is no content

This is a preparatory GUI cleanup, prior to the addition of further properties in additional tabs.

Before:
![grafik](https://github.com/user-attachments/assets/93ca7d75-d91e-47e7-9509-5c752332cf6b)

This PR:
![grafik](https://github.com/user-attachments/assets/f3089b98-2b99-4d57-bace-406caefa80b2)

JoergAtGithub committed:
Convert groupBoxSettings/groupBoxScreens into tabs
Show only the tabs, which have valid content in the current context
Sort tabs that settingsTab is shown by default, if controller settings are defined in mapping
Moved controller tabs below info and warning section
Made hyperlink warning label clickable and warning text select- and copy-able
Made warning label collapsing in height
Hide 'Mapping Info' fields if there is no content

Swiftb0y commented on src/controllers/dlgprefcontroller.h:
```diff
@@ -140,4 +137,6 @@ class DlgPrefController : public DlgPreferencePage {
     ControllerMappingTableProxyModel* m_pOutputProxyModel;
     bool m_GuiInitialized;
     bool m_bDirty;
+    int m_inputMappingsTabIndex;  // Index of the input mappings tab
+    int m_outputMappingsTabIndex; // Index of the output mappings tab
```
```suggestion
    int m_inputMappingsTabIndex;  // Index of the input mappings tab in m_ui.controllerTabs
    int m_outputMappingsTabIndex; // Index of the output mappings tab in m_ui.controllerTabs
```

Swiftb0y reviewed (commented):
LGTM otherwise. Thank you.

Swiftb0y reviewed (approved):
LGTM. Waiting for CI. @ronso0 do you have interest to take a quick look at the `.ui` changes?

ronso0 commented:
Umm, looking at `.ui` diffs is pointless here. Before and after grid/layout items are in random order (that's why I usually reorder them so GUI order matches .ui order).
And I can't built main currently because I don't have hidapi 0.14.0

So if the screenshots reflect the current state of this PR and UX (correct tabstops etc.) is the same this LGTM :+1: 
Just the grids/2-column-layouts in Device Info / Mapping Info could be optimised so that eg. Mapping -> Description can use the available space.

JoergAtGithub committed:
Reordered widgets according position in ui grid
Added tabstop section
Unified some names
Simplified top level frame/grid structure to a single grid

JoergAtGithub commented:
I reordered the .ui file in GUI order.
I added a tabstop section.
But I could not find a way to use all the available vertical space in the grids/2-column-layouts in Device Info / Mapping Info. I tried for hours, and the only partially working solution was to set the spacer height from C++, which I consider more ugly than a little uneccessary whitespace.

ronso0 commented:
I'll take a look asap, can build again with a libhidapi backport.

ronso0 commented on src/controllers/dlgprefcontroller.cpp:
```diff
@@ -992,25 +988,51 @@ void DlgPrefController::slotShowPreviewScreens(
+    // because without them the layout would be collapse.
+
+    // Remove all tabs first
+    m_ui.controllerTabs->clear();
```
Why remove and insert tabs?
There is also [`QTabWidget::setTabVisible(int index, bool visible)`](https://doc.qt.io/qt-6/qtabwidget.html#setTabVisible)
IIUC tab indices are unchanged (unlike visual position in ui) so we'd only need to store them in the ctor.

This way we'd also get around double tab tr strings (here and in ui file)

ronso0 replied:
same applies to the screens and settings tab, of course.

ronso0 commented on src/controllers/dlgprefcontroller.cpp:
```diff
@@ -1019,13 +1041,15 @@ void DlgPrefController::slotShowMapping(std::shared_ptr<LegacyControllerMapping>
+    // Show or hide the settings tab based on the presence of settings
+    if (pMapping && !pMapping->getSettings().isEmpty()) {
+        if (m_ui.controllerTabs->indexOf(m_ui.settingsTab) == -1) {
+            m_ui.controllerTabs->addTab(m_ui.settingsTab, tr("Mapping Settings"));
```
This is obsolete with `settabVisible()`, so just FYI to avoid duplicate tr strings here and n ui file, we could use `QTabwidget::tabText(QTabwidget::indexOf(QWidget*))`

ronso0 reviewed (changes requested):
my first review, will take another look soonish

ronso0 commented:
A regression I noticed is that now the Input/Output tabs can't use the entire height of the dialog as before, mapping and device info will always occupy space at the top. Also 

One solution would be to allow collapsing groupboxes.
1. subclass QGroupBox, add collapse/expand buttons, which I'd appreciate also for mappings with many options where I'm interested only in one or two settings catgeories.
Not trivial though..
2. add an expand/collapse toggle above Device / Mapping Info.
Much easier, but the toggle itself would occupy a row

ronso0 commented:
Btw
QCheckBox "chkEnabledDevice" and QLabel "labelLoadMapping" are both in row="1" column="0".
This will probably never cause issues (overlap) because the checkbox is left-aligned (default) and the label is right-aligned, only if the entire dialog is very narrow **and** all device info fields are empty **and** the big, bold device name is very short **and** the mapping description is not empty.
Just saying..

JoergAtGithub commented:
> Btw QCheckBox "chkEnabledDevice" and QLabel "labelLoadMapping" are both in row="1" column="0".

I also noticed this, but this is not new in this PR.

JoergAtGithub committed:
Replaced dynamic tab addition/removal with setTabVisible.
Added methods to count and find the first visible tab.

ronso0 reviewed (commented):
Thanks, I'll do a test with some HID and MIDI gear... next year ; )

ronso0 commented on src/controllers/dlgprefcontroller.cpp:
```diff
@@ -796,6 +776,39 @@ void DlgPrefController::slotMappingSelected(int chosenIndex) {
     }
+
+    // These tabs are only usable for MIDI controllers
+    if (m_pController->getDataRepresentationProtocol() == DataRepresentationProtocol::MIDI &&
+            !mappingFilePath.isEmpty()) {
+        m_ui.controllerTabs->setTabVisible(m_inputMappingsTabIndex, true);
+        m_ui.controllerTabs->setTabVisible(m_outputMappingsTabIndex, true);
+
+    } else {
+        m_ui.controllerTabs->setTabVisible(m_inputMappingsTabIndex, false);
+        m_ui.controllerTabs->setTabVisible(m_outputMappingsTabIndex, false);
+    }
```
just an idea..
```suggestion
    bool showIOtabs = m_pController->getDataRepresentationProtocol() == DataRepresentationProtocol::MIDI &&
            !mappingFilePath.isEmpty();
    m_ui.controllerTabs->setTabVisible(m_inputMappingsTabIndex, showIOtabs);
    m_ui.controllerTabs->setTabVisible(m_outputMappingsTabIndex, showIOtabs);
```

JoergAtGithub committed:
Simplified the logic for showing/hiding MIDI controller tabs by introducing a boolean variable

JoergAtGithub commented:
> Thanks, I'll do a test with some HID and MIDI gear... next year ; )

@ronso0 Did you already found the time to test this?

ronso0 commented:
Sorry, this got shifted down in my TODO list.
Will try to test tonight.

ronso0 committed:
DlgPrefController ui: remove useless rowstretch (ignored if all are 0)

ronso0 committed:
DlgPrefController ui: remove useless minimumSize from mapping description

ronso0 committed:
DlgPrefController ui: keep device info small, expand mapping info horizontally

ronso0 committed:
DlgPrefController ui: keep device/mapping info minimal vertically

ronso0 committed:
DlgPrefController ui: remove "handling" from data protocol label

ronso0 committed:
DlgPrefController: remove screens tab if QML flag is OFF

ronso0 committed:
DlgPrefController: rename function, remove blank line

ronso0 commented on src/controllers/dlgprefcontroller.cpp:
```diff
@@ -796,6 +776,35 @@ void DlgPrefController::slotMappingSelected(int chosenIndex) {
+    return visibleTabsCount;
+}
+
+int DlgPrefController::getIndexOfFirstVisibleTabs() {
```
fixed in https://github.com/JoergAtGithub/mixxx/pull/3

ronso0 commented on src/controllers/dlgprefcontroller.cpp:
```diff
@@ -296,6 +305,12 @@ DlgPrefController::DlgPrefController(
             &QPushButton::clicked,
             this,
             &DlgPrefController::slotOutputControlSearch);
+
+    // Store the index of the input and output mappings tabs
+    m_inputMappingsTabIndex = m_ui.controllerTabs->indexOf(m_ui.inputMappingsTab);
+    m_outputMappingsTabIndex = m_ui.controllerTabs->indexOf(m_ui.outputMappingsTab);
+    m_settingsTabIndex = m_ui.controllerTabs->indexOf(m_ui.settingsTab);
+    m_screensTabIndex = m_ui.controllerTabs->indexOf(m_ui.screensTab);
```
I don't see any side effects, just saying..

ronso0 reviewed (commented):
LGTM otherwise!

I opened https://github.com/JoergAtGithub/mixxx/pull/3 with some fixes:
* Screens tab was always visible when built with QML off
* finally managed to fix the random layout changes / v-stretch of the device / mapping info groupboxes
(this took hours, literally... sometimes Qt widget/layout props are a PITA)
* nits

Tested with MIDI Through and Traktor S4mk3 (HID).
All looking good now with the fixup PR.
![image](https://github.com/user-attachments/assets/fa4fab15-e115-421e-b1d0-0f64497c3cfc)

JoergAtGithub committed:
Merge pull request #3 from ronso0/con-pref-tab-layout-fixes

DlgPrefController layout fixes

JoergAtGithub commented on src/controllers/dlgprefcontroller.cpp:
```diff
@@ -1107,6 +1136,15 @@ void DlgPrefController::slotShowMapping(std::shared_ptr<LegacyControllerMapping>
+    int firstVisibleTabIndex = getIndexOfFirstVisibleTabs();
+    if (firstVisibleTabIndex >= 0) {
+        m_ui.controllerTabs->setCurrentIndex(firstVisibleTabIndex);
+    }
```
I see your use case. But I think another use case is more important:
If a (new) user steps/scrolls through the list of mappings. Than it's likely that there was a mapping without controller settings beneth them. This means Midi In becomes the default - an expert setting, not intendend for average users. For them we introduced the controller settings, which should appear as default therefore.

ronso0 commented:
Okay, let's merge it then.
Thanks!
